Index: dnsmasq-2.85/src/dhcp-common.c
===================================================================
--- dnsmasq-2.85.orig/src/dhcp-common.c	2021-04-21 17:17:03.590314462 +0800
+++ dnsmasq-2.85/src/dhcp-common.c	2021-04-21 16:50:49.758368600 +0800
@@ -646,6 +646,7 @@
   "classless-static-route", 121, 0}, {
   "vendor-id-encap", 125, 0}, {
   "tftp-server-address", 150, OT_ADDR_LIST}, {
+  "captive-portal", 160, OT_NAME}, {
   "server-ip-address", 255, OT_ADDR_LIST},      /* special, internal only, sets siaddr */
   {
   NULL, 0, 0}
Index: dnsmasq-2.85/src/dhcp-protocol.h
===================================================================
--- dnsmasq-2.85.orig/src/dhcp-protocol.h	2021-04-21 17:17:03.590314462 +0800
+++ dnsmasq-2.85/src/dhcp-protocol.h	2021-04-21 16:51:16.840283900 +0800
@@ -64,6 +64,7 @@
 #define OPTION_SIP_SERVER        120
 #define OPTION_VENDOR_IDENT      124
 #define OPTION_VENDOR_IDENT_OPT  125
+#define OPTION_CAPTIVE_PORTAL    160
 #define OPTION_END               255
 
 #define SUBOPT_CIRCUIT_ID        1
