Index: dnsmasq-2.85/src/cache.c
===================================================================
--- dnsmasq-2.85.orig/src/cache.c	2021-04-22 11:20:04.924439806 +0800
+++ dnsmasq-2.85/src/cache.c	2021-04-22 10:55:28.690776600 +0800
@@ -643,6 +643,9 @@
 }
 
 #ifdef DNS_TRAFFIC_ROUTE
+#if defined(ZYXEL_DNS_TRAFFIC_ROUTE_FUZZY_COMPARE)
+#define DNS_ROUTE_FUZZY_MATCH_HEADER	"*."
+#endif
 void add_dns_traffic_route(uint32_t s_addr, const char *cname, int *flush)
 {
   char sysCmd[128] = "";
@@ -650,6 +653,13 @@
   srcIP.s_addr = s_addr;
   struct server *serv = NULL;
 
+#if defined(ZYXEL_DNS_TRAFFIC_ROUTE_FUZZY_COMPARE)
+  char *pFuzzyName = NULL, *pRemoveHeaderCName = NULL, cTmpCName[MAXDNAME];
+  uint8_t i, uCNameLength, uMatch = 0, uFuzzyLengh = strlen (DNS_ROUTE_FUZZY_MATCH_HEADER);
+#else
+  uint8_t uMatch = 0;
+#endif //#if defined(ZYXEL_DNS_TRAFFIC_ROUTE_FUZZY_COMPARE)
+
   if (s_addr == 0) {
 
     return;
@@ -658,7 +668,35 @@
   serv = daemon->servers;
 
   while (serv) {
+#if defined(ZYXEL_DNS_TRAFFIC_ROUTE_FUZZY_COMPARE)
+    if (serv->domain) {
+      if (strncmp (serv->domain, DNS_ROUTE_FUZZY_MATCH_HEADER, uFuzzyLengh) == 0) {
+        pFuzzyName = serv->domain + uFuzzyLengh;
+
+        pRemoveHeaderCName = NULL;
+        strncpy (cTmpCName, cname, MAXDNAME);
+        uCNameLength = strlen (cTmpCName);
+
+        for (i = 0 ; i < uCNameLength ; i++) {
+          if (cTmpCName [i] == '.') {
+            pRemoveHeaderCName = (cTmpCName + (i + 1));
+            break;
+          }
+        }
+
+        if (pRemoveHeaderCName != NULL && strcmp (pFuzzyName, pRemoveHeaderCName) == 0) {
+          uMatch = 1;
+        }
+      } else if (strcmp(cname, serv->domain) == 0) {
+        uMatch = 1;
+      }
+    }
+#else
     if (serv->domain && (strcmp(cname, serv->domain) == 0)) {
+      uMatch = 1;
+    }
+#endif //#if defined(ZYXEL_DNS_TRAFFIC_ROUTE_FUZZY_COMPARE)
+    if (uMatch == 1) {
       if (*flush) {
         /* deledt */
         sprintf(sysCmd,
