Index: dnsmasq-2.85/src/dns-protocol.h
===================================================================
--- dnsmasq-2.85.orig/src/dns-protocol.h
+++ dnsmasq-2.85/src/dns-protocol.h
@@ -78,11 +78,11 @@
 #define T_ANY		255
 #define T_CAA           257
 
-#ifdef ZYXEL_TTNET_EDNS //add ppp username to edns option for TT
-#define EDNS0_OPTION_USERNAME       65001 /* Nominum temporary assignment *///according to TT request set it to 65001
+#ifdef ZYXEL_TAAAG_EDNS //add ppp username to edns option for TAAAB
+#define EDNS0_OPTION_USERNAME       65001 /* Nominum temporary assignment *///according to TAAAB request set it to 65001
 #define EDNS0_MAC_HEAD 16389
 #define EDNS0_USERNAME_HEAD 32 
-#define EDNS0_OPTION_MAC            38       /* dyndns.org temporary assignment *///according to TT request set it to 38
+#define EDNS0_OPTION_MAC            38       /* dyndns.org temporary assignment *///according to TAAAB request set it to 38
 #else
 #define EDNS0_OPTION_MAC            65001       /* dyndns.org temporary assignment */
 #endif
Index: dnsmasq-2.85/src/dnsmasq.h
===================================================================
--- dnsmasq-2.85.orig/src/dnsmasq.h
+++ dnsmasq-2.85/src/dnsmasq.h
@@ -287,7 +287,7 @@ struct event_desc
 #define MS_SCRIPT LOG_MAIL
 #define MS_DEBUG  LOG_NEWS
 
-#ifdef CONFIG_TT_CUSTOMIZATION_TR069
+#ifdef CONFIG_TAAAB_CUSTOMIZATION_TR069
 struct query_domain {
   char domainname[MAXDNAME];
   int queryCount;
@@ -1174,7 +1174,7 @@ extern struct daemon
   unsigned long local_ttl, neg_ttl, max_ttl, min_cache_ttl, max_cache_ttl,
     auth_ttl, dhcp_ttl, use_dhcp_ttl;
   char *dns_client_id;
-#ifdef ZYXEL_TTNET_EDNS// add ppp username to edns option for TT
+#ifdef ZYXEL_TAAAG_EDNS// add ppp username to edns option for TAAAB
   char *dns_ppp_username;
 #endif
   struct hostsfile *addn_hosts;
@@ -1248,7 +1248,7 @@ extern struct daemon
   int pipe_to_parent;
   int numrrand;
   struct randfd *randomsocks;
-#ifdef CONFIG_TT_CUSTOMIZATION_TR069
+#ifdef CONFIG_TAAAB_CUSTOMIZATION_TR069
   struct query_domain *domainQuery;
 #endif
   struct randfd_list *rfl_spare, *rfl_poll;
Index: dnsmasq-2.85/src/edns0.c
===================================================================
--- dnsmasq-2.85.orig/src/edns0.c
+++ dnsmasq-2.85/src/edns0.c
@@ -102,7 +102,7 @@ add_pseudoheader (struct dns_header * he
   int rdlen = 0, is_sign, is_last;
   unsigned short flags = set_do ? 0x8000 : 0, rcode = 0;
 
-#ifdef ZYXEL_TTNET_EDNS//  add ppp username to edns option for TT
+#ifdef ZYXEL_TAAAG_EDNS//  add ppp username to edns option for TAAAB
   int ttnet_edns0_opt_head = 0 ;
 #endif
   p = find_pseudoheader (header, plen, NULL, &udp_len, &is_sign, &is_last);
@@ -225,7 +225,7 @@ add_pseudoheader (struct dns_header * he
     if (p + 4 > limit)
       return plen;              /* Too big */
     PUTSHORT (optno, p);
-#ifdef ZYXEL_TTNET_EDNS//  add ppp username to edns option for TT
+#ifdef ZYXEL_TAAAG_EDNS//  add ppp username to edns option for TAAAB
 	if(optno == EDNS0_OPTION_MAC)
 	  ttnet_edns0_opt_head = EDNS0_MAC_HEAD ;
 	else if(optno == EDNS0_OPTION_USERNAME)
@@ -307,7 +307,7 @@ add_mac (struct dns_header *header, size
 
   if ((maclen = find_mac (l3, mac, 1, now)) != 0) {
     *cacheablep = 0;
-#ifdef ZYXEL_TTNET_EDNS//  add ppp username to edns option for TT
+#ifdef ZYXEL_TAAAG_EDNS//  add ppp username to edns option for TAAAB
 	if (daemon->dns_ppp_username)
 	  plen = add_pseudoheader(header, plen, limit, PACKETSZ, EDNS0_OPTION_USERNAME, 
 				(unsigned char *)daemon->dns_ppp_username, strlen(daemon->dns_ppp_username), 0, 0);
Index: dnsmasq-2.85/src/forward.c
===================================================================
--- dnsmasq-2.85.orig/src/forward.c
+++ dnsmasq-2.85/src/forward.c
@@ -237,7 +237,7 @@ search_servers (time_t now, union all_ad
   return flags;
 }
 
-#ifdef CONFIG_TT_CUSTOMIZATION_TR069
+#ifdef CONFIG_TAAAB_CUSTOMIZATION_TR069
 static struct query_domain *
 create_domain_rec (char *namebuff, unsigned int QType)
 {
@@ -335,7 +335,7 @@ forward_query (int udpfd, union mysockad
 #endif
   unsigned int gotname =
     extract_request (header, plen, daemon->namebuff, NULL);
-#ifdef CONFIG_TT_CUSTOMIZATION_TR069
+#ifdef CONFIG_TAAAB_CUSTOMIZATION_TR069
   int i = 0;
   int forwardFlag = 0;          //richard
   time_t now_time;
@@ -543,7 +543,7 @@ forward_query (int udpfd, union mysockad
     size_t edns0_len;
     unsigned char *pheader;
 
-#ifdef CONFIG_TT_CUSTOMIZATION_TR069
+#ifdef CONFIG_TAAAB_CUSTOMIZATION_TR069
     struct query_domain *n = NULL;
     int found = 0;
 
@@ -703,7 +703,7 @@ forward_query (int udpfd, union mysockad
           log_query_mysockaddr (F_SERVER | F_FORWARD, daemon->namebuff,
                                 &start->addr, NULL);
           start->queries++;
-#ifdef CONFIG_TT_CUSTOMIZATION_TR069
+#ifdef CONFIG_TAAAB_CUSTOMIZATION_TR069
           if (n != NULL) {
             n->queryCount++;
           }
@@ -721,7 +721,7 @@ forward_query (int udpfd, union mysockad
               forward->forwardall = 1;
           }
 #endif
-#ifndef CONFIG_TT_CUSTOMIZATION_TR069
+#ifndef CONFIG_TAAAB_CUSTOMIZATION_TR069
           if (!forward->forwardall)
 #endif
             break;
Index: dnsmasq-2.85/src/option.c
===================================================================
--- dnsmasq-2.85.orig/src/option.c
+++ dnsmasq-2.85/src/option.c
@@ -177,7 +177,7 @@ struct myoption
 #define LOPT_TAG_IFNAME   364
 #define LOPT_VENDOREXACT  365
 #endif
-#ifdef ZYXEL_TTNET_EDNS// add ppp username to edns option for TT
+#ifdef ZYXEL_TAAAG_EDNS// add ppp username to edns option for TAAAB
 #define LOPT_PPP_USER      366
 #endif
 
@@ -320,7 +320,7 @@ static const struct myoption opts[] =
   {"add-mac", 2, 0, LOPT_ADD_MAC},
   {"add-subnet", 2, 0, LOPT_ADD_SBNET},
   {"add-cpe-id", 1, 0, LOPT_CPE_ID},
-#ifdef ZYXEL_TTNET_EDNS// add ppp username to edns option for TT
+#ifdef ZYXEL_TAAAG_EDNS// add ppp username to edns option for TAAAB
   { "add-ppp-username", 1, 0 , LOPT_PPP_USER },
 #endif
   {"proxy-dnssec", 0, 0, LOPT_DNSSEC},
@@ -717,7 +717,7 @@ static struct
   LOPT_CPE_ID, ARG_ONE, "<text>",
       gettext_noop ("Add client identification to forwarded DNS queries."),
       NULL}, {
-#ifdef ZYXEL_TTNET_EDNS// add ppp username to edns option for TT
+#ifdef ZYXEL_TAAAG_EDNS// add ppp username to edns option for TAAAB
   LOPT_PPP_USER, ARG_ONE, "<text>", 
   		gettext_noop("Add CPE PPPoE username to forwarded DNS queries."), 
   		NULL },{
@@ -2744,7 +2744,7 @@ one_opt (int option, char *arg, char *er
     if (arg)
       daemon->dns_client_id = opt_string_alloc (arg);
     break;
-#ifdef ZYXEL_TTNET_EDNS// add ppp username to edns option for TT
+#ifdef ZYXEL_TAAAG_EDNS// add ppp username to edns option for TAAAB
 		case LOPT_PPP_USER: /* --add-ppp-username */
 			if(arg)
 				daemon->dns_ppp_username = opt_string_alloc(arg);
