diff -urN clinkc.org/include/cybergarage/http/chttp.h clinkc/include/cybergarage/http/chttp.h
--- clinkc.org/include/cybergarage/http/chttp.h	2021-01-25 15:02:27.982077400 +0800
+++ clinkc/include/cybergarage/http/chttp.h	2021-01-26 13:35:55.015483462 +0800
@@ -71,7 +71,7 @@
 #define BOUNDARY_SIZE 256
 #define HTTPD_TMP_PATH "/tmp"
 //#define HTTPD_MAX_UPLOAD_SIZE 41943040
-#define CSRFTOKEN_SIZE 64
+#define CSRFTOKEN_BUFFER_SIZE 65
 #define HTTPD_CSRFTOKEN "CSRFToken"
 #endif
 
@@ -360,7 +360,7 @@
 	char loginUserName[32];
 	char loginLevel[16];
 	char sessionID[32+1];
-	char CSRFToken[CSRFTOKEN_SIZE];
+	char CSRFToken[CSRFTOKEN_BUFFER_SIZE];
 	BOOL needChgPasswd;
 	BOOL showSkipBtn; /* show skip button in change password page if password is automatically generated */
 	BOOL needQuickStart;
diff -urN clinkc.org/src/cybergarage/http/chttp_request.c clinkc/src/cybergarage/http/chttp_request.c
--- clinkc.org/src/cybergarage/http/chttp_request.c	2021-01-25 15:02:27.982077400 +0800
+++ clinkc/src/cybergarage/http/chttp_request.c	2021-01-26 13:36:06.947120745 +0800
@@ -905,7 +905,8 @@
 	}
 
 	if(tokenvalue = cg_http_packet_getheadervalue((CgHttpPacket *)httpReq, HTTPD_CSRFTOKEN)){
-		strncpy(httpReq->CSRFToken, tokenvalue, CSRFTOKEN_SIZE);
+		memset(httpReq->CSRFToken, '\0', CSRFTOKEN_BUFFER_SIZE);
+		memcpy(httpReq->CSRFToken, tokenvalue, CSRFTOKEN_BUFFER_SIZE);
 	}
 	if (failed == TRUE) return FALSE;
 #if 0
