Index: ppp-2.4.3/pppd/options.c
===================================================================
--- ppp-2.4.3.orig/pppd/options.c	2021-09-29 20:02:56.524893400 +0800
+++ ppp-2.4.3/pppd/options.c	2021-10-05 14:01:39.286817416 +0800
@@ -1307,9 +1307,10 @@
 	    /*
 	     * Store the resulting character for the escape sequence.
 	     */
-	    if (len < MAXWORDLEN-1)
+	    if (len < MAXWORDLEN) {
 		word[len] = value;
-	    ++len;
+	      ++len;
+	    }
 
 	    if (!got)
 		c = getc(f);
@@ -1342,9 +1343,10 @@
 	/*
 	 * An ordinary character: store it in the word and get another.
 	 */
-	if (len < MAXWORDLEN-1)
+	if (len < MAXWORDLEN) {
 	    word[len] = c;
-	++len;
+	    ++len;
+	}
 
 	c = getc(f);
     }