Index: radvd-1.8/includes.h
===================================================================
--- radvd-1.8.orig/includes.h
+++ radvd-1.8/includes.h
@@ -73,7 +73,12 @@
 
 #include <arpa/inet.h>
 
+
+#ifndef LIBC_USE_MUSL
 #include <sys/sysctl.h>
+#else
+#include <linux/sysctl.h>
+#endif
 
 #include <net/if.h>
 
Index: radvd-1.8/radvd.c
===================================================================
--- radvd-1.8.orig/radvd.c
+++ radvd-1.8/radvd.c
@@ -20,6 +20,12 @@
 #include "radvd.h"
 #include "pathnames.h"
 
+#ifdef LIBC_USE_MUSL
+#define _GNU_SOURCE
+#include <sys/syscall.h>
+#include <linux/sysctl.h>
+#endif
+
 #ifdef HAVE_NETLINK
 #include "netlink.h"
 #endif
@@ -888,8 +894,13 @@ check_ip6_forwarding(void)
 	                        "or the kernel interface has changed?");
 #endif /* __linux__ */
 
+#ifndef LIBC_USE_MUSL
 	if (!fp && sysctl(forw_sysctl, sizeof(forw_sysctl)/sizeof(forw_sysctl[0]),
 	    &value, &size, NULL, 0) < 0) {
+#else
+	//if (!fp && syscall(SYS__sysctl, NULL) < 0) {
+	if (!fp) {
+#endif
 		flog(LOG_DEBUG, "Correct IPv6 forwarding sysctl branch not found, "
 			"perhaps the kernel interface has changed?");
 		return(0);	/* this is of advisory value only */
@@ -1136,7 +1148,12 @@ void set_prefix_to_old(void)
 				// write into prefix
 				int i=0;
 				for(;i<16;i++){
+
+#ifndef LIBC_USE_MUSL
 					prefix->Prefix.__in6_u.__u6_addr8[i] = buffer[i];
+#else
+					prefix->Prefix.__in6_union.__s6_addr[i] = buffer[i];
+#endif
 				}
 				free(buffer);
 				fclose(pFile);
@@ -1182,7 +1199,12 @@ void set_prefix_back_to_new(void)
 				// write into prefix
 				int i=0;
 				for(;i<16;i++){
+
+#ifndef LIBC_USE_MUSL
 					prefix->Prefix.__in6_u.__u6_addr8[i] = buffer[i];
+#else
+					prefix->Prefix.__in6_union.__s6_addr[i] = buffer[i];
+#endif
 				}
 				free(buffer);
 				fclose(pFile);
@@ -1272,7 +1294,7 @@ void backup_old_prefix(void)
             unsigned char buff[16];
             int i=0;
             for(;i<16;i++){
-                buff[i] = prefix->Prefix.__in6_u.__u6_addr8[i];
+                buff[i] = prefix->Prefix.__in6_union.__s6_addr[i];
             }
             FILE* fp = NULL;
             fp = fopen("/tmp/radvd_oldprefix", "wb");
@@ -1309,7 +1331,12 @@ void backup_new_prefix(void)
             unsigned char buff[16];
             int i=0;
             for(;i<16;i++){
+
+#ifndef LIBC_USE_MUSL
                 buff[i] = prefix->Prefix.__in6_u.__u6_addr8[i];
+#else
+                buff[i] = prefix->Prefix.__in6_union.__s6_addr[i];
+#endif
             }
             FILE* fp = NULL;
             fp = fopen("/tmp/radvd_newprefix", "wb");
