Index: rpcd-2021-03-11-ccb75178/plugin.c
===================================================================
--- rpcd-2021-03-11-ccb75178.orig/plugin.c
+++ rpcd-2021-03-11-ccb75178/plugin.c
@@ -100,6 +100,10 @@ rpc_plugin_json_element_to_blob(const ch
 	case json_type_null:
 		blobmsg_add_field(blob, BLOBMSG_TYPE_UNSPEC, name, NULL, 0);
 		break;
+		
+	default:
+		blobmsg_add_field(blob, BLOBMSG_TYPE_UNSPEC, name, NULL, 0);
+		break;
 	}
 }
 
