Index: linux-4.19.183/net/netfilter/nf_conntrack_proto_tcp.c
===================================================================
--- linux-4.19.183.orig/net/netfilter/nf_conntrack_proto_tcp.c
+++ linux-4.19.183/net/netfilter/nf_conntrack_proto_tcp.c
@@ -37,7 +37,7 @@
 /* "Be conservative in what you do,
     be liberal in what you accept from others."
     If it's non-zero, we mark only out of window RST segments as INVALID. */
-static int nf_ct_tcp_be_liberal __read_mostly = 0;
+static int nf_ct_tcp_be_liberal __read_mostly = 1;
 
 /* If it is set to zero, we disable picking up already established
    connections. */
