Index: linux-4.19.151/net/netfilter/nf_conntrack_helper.c
===================================================================
--- linux-4.19.151.orig/net/netfilter/nf_conntrack_helper.c
+++ linux-4.19.151/net/netfilter/nf_conntrack_helper.c
@@ -37,7 +37,7 @@ unsigned int nf_ct_helper_hsize __read_m
 EXPORT_SYMBOL_GPL(nf_ct_helper_hsize);
 static unsigned int nf_ct_helper_count __read_mostly;
 
-static bool nf_ct_auto_assign_helper __read_mostly = false;
+static bool nf_ct_auto_assign_helper __read_mostly = true;
 module_param_named(nf_conntrack_helper, nf_ct_auto_assign_helper, bool, 0644);
 MODULE_PARM_DESC(nf_conntrack_helper,
 		 "Enable automatic conntrack helper assignment (default 0)");
