Index: xl2tpd-1.3.14/xl2tpd.c
===================================================================
--- xl2tpd-1.3.14.orig/xl2tpd.c	2019-11-08 02:58:09.499933466 +0000
+++ xl2tpd-1.3.14/xl2tpd.c	2019-11-08 02:59:09.336208679 +0000
@@ -1880,6 +1880,9 @@
     signal (SIGUSR1, &sigusr1_handler);
     signal (SIGHUP, &sighup_handler);
     signal (SIGPIPE, SIG_IGN);
+
+    // ignore SIGTTOU for tcsetattr. tcsetattr called by background process will send SIGTTOU signal to process group
+    signal (SIGTTOU, SIG_IGN);
     init_scheduler ();
 
     unlink(gconfig.controlfile);
