Index: busybox-1.31.1/init/init.c
===================================================================
--- busybox-1.31.1.orig/init/init.c	2021-03-31 19:45:16.132131947 +0800
+++ busybox-1.31.1/init/init.c	2021-03-31 19:22:23.134570800 +0800
@@ -236,6 +236,7 @@
 #define FLASH_LOG_FILE_PATH "/data/"
 #define FLASH_LOG_FILE_SYSLOG "syslog.log"
 #define FLASH_LOG_FILE_SECURITY "security.log"
+#define FLASH_LOG_FILE_NAT "nat.log"
 #define LOG_ROTATE 2
 
 void rotateLog(char *flashLogFilePath, int logFileRotate);
@@ -1037,6 +1038,11 @@
 		sprintf(tmpfilename,"%s%s",LOG_FILE_PATH, FLASH_LOG_FILE_SECURITY);
 		sprintf(tmpflashfilename,"%s%s",FLASH_LOG_FILE_PATH, FLASH_LOG_FILE_SECURITY);
 		copyLogToFlash(tmpfilename, tmpflashfilename, 200*1024, LOG_ROTATE);
+#ifdef ZYXEL_NAT_LOG
+		sprintf(tmpfilename,"%s%s",LOG_FILE_PATH, FLASH_LOG_FILE_NAT);
+		sprintf(tmpflashfilename,"%s%s",FLASH_LOG_FILE_PATH, FLASH_LOG_FILE_NAT);
+		copyLogToFlash(tmpfilename, tmpflashfilename, 200*1024, LOG_ROTATE);
+#endif
 	}
 #endif
 
