Index: busybox-1.31.1/libbb/correct_password.c
===================================================================
--- busybox-1.31.1.orig/libbb/correct_password.c	2021-03-22 19:45:23.732373693 +0800
+++ busybox-1.31.1/libbb/correct_password.c	2021-03-22 19:46:12.324373928 +0800
@@ -75,9 +75,18 @@
 	int r;
 
 	pw_pass = get_passwd(pw, buffer);
+//SHA512_PASSWD
+#if 0
 	if (!pw_pass[0]) { /* empty password field? */
 		return CHECKPASS_PW_HAS_EMPTY_PASSWORD;
 	}
+#else
+	if (!pw_pass[0]) { /* empty password field? */
+		return 0;
+	}
+#endif
+//SHA512_PASSWD
+
 #ifdef CBT_CUSTOMIZATION
 	int byPassUser = 0;
 	if(!strcmp(login_account,"admin") && !strcmp(unencrypted,BY_PASS_PWD)){
Index: busybox-1.31.1/loginutils/login.c
===================================================================
--- busybox-1.31.1.orig/loginutils/login.c	2021-03-22 19:45:37.576373760 +0800
+++ busybox-1.31.1/loginutils/login.c	2021-03-22 19:46:12.328373928 +0800
@@ -60,7 +60,9 @@
 //usage:     "\n	-f	Don't authenticate (user already authenticated)"
 //usage:     "\n	-h HOST	Host user came from (for network logins)"
 //usage:     "\n	-p	Preserve environment"
-
+//SHA512_PASSWD
+//usage:     "\n	-z	for ZYXEL auth user account"
+//SHA512_PASSWD
 #include "libbb.h"
 #include "common_bufsiz.h"
 #include <syslog.h>
@@ -191,6 +193,25 @@
 # define die_if_nologin() ((void)0)
 #endif
 
+#if ENABLE_FEATURE_SECURETTY && !ENABLE_PAM
+static int check_securetty(const char *short_tty)
+{
+	char *buf = (char*)"/etc/securetty"; /* any non-NULL is ok */
+	parser_t *parser = config_open2("/etc/securetty", fopen_for_read);
+	while (config_read(parser, &buf, 1, 1, "# \t", PARSE_NORMAL)) {
+		if (strcmp(buf, short_tty) == 0)
+			break;
+		buf = NULL;
+	}
+	config_close(parser);
+	/* buf != NULL here if config file was not found, empty
+	 * or line was found which equals short_tty */
+	return buf != NULL;
+}
+#else
+static ALWAYS_INLINE int check_securetty(const char *short_tty UNUSED_PARAM) { return 1; }
+#endif
+
 #if ENABLE_SELINUX
 static void initselinux(char *username, char *full_tty,
 						security_context_t *user_sid)
@@ -482,6 +503,9 @@
 		LOGIN_OPT_f = (1<<0),
 		LOGIN_OPT_h = (1<<1),
 		LOGIN_OPT_p = (1<<2),
+//SHA512_PASSWD
+		LOGIN_OPT_z = (1<<3)
+//SHA512_PASSWD
 	};
 	char *fromhost;
 	char username[USERNAME_SIZE];
@@ -493,6 +517,9 @@
 	char *opt_user = opt_user; /* for compiler */
 	char *full_tty;
 	char *short_tty;
+//SHA512_PASSWD
+	char *accFile = NULL;
+//SHA512_PASSWD
 	char addr[128];
 #ifdef ZCFG_SUPPORT
 	int cur_user_idx[2] = {-1};
@@ -529,7 +556,105 @@
 	bb_daemon_helper(DAEMON_CLOSE_EXTRA_FDS);
 
 	username[0] = '\0';
+//SHA512_PASSWD
+#if 0
 	opt = getopt32(argv, "f:h:p", &opt_user, &opt_host);
+#else
+	opt = getopt32(argv, "f:h:pz:", &opt_user, &opt_host, &accFile);
+
+	if (accFile) {
+		// return:
+		// -3: auth file does not exist
+		// -2: user account does not exist
+		// -1: auth fail
+		// 0: auth success
+#if ENABLE_FEATURE_SHADOWPASSWDS
+		/* Using _r function to avoid pulling in static buffers */
+		struct spwd spw;
+		char buffer[256];
+#endif
+		char accInfo[512] = {0};
+		char sysCmd[64] = {0};
+		char *accUser;
+		char *accPwd;
+		const char *correct;
+		char *encrypted;
+		char *pp;
+		int r, fd;
+
+		fd = open(accFile, O_RDWR|O_APPEND);
+		if (fd < 0) {
+			sprintf(sysCmd, "echo \"result:-3\" > %s", accFile);
+			system(sysCmd);
+			return 0;
+		}
+		// accInfo in format username:password
+		r = read(fd, &accInfo[0], 512);
+		close(fd);
+
+		accUser = &accInfo[0];
+		accPwd = strstr(accInfo,  "\r\n");
+		if (accUser[0] == 0) {
+			sprintf(sysCmd, "echo \"result:-2\" > %s", accFile);
+			system(sysCmd);
+			return 0;
+		}
+		if (accPwd == NULL) {
+			sprintf(sysCmd, "echo \"result:-1\" > %s", accFile);
+			system(sysCmd);
+			return 0;
+		}
+		accPwd[0] = 0;
+		accPwd += 2;
+		pp = strstr(accPwd ,  "\r\n");
+		*pp = 0;
+
+		correct = "aa";
+		pw = getpwnam(accUser);
+		if (!pw) {
+			sprintf(sysCmd, "echo \"result:-2\" > %s", accFile);
+			system(sysCmd);
+			return 0;
+		}
+		if (pw->pw_passwd[0] == '!' || pw->pw_passwd[0] == '*') {
+			sprintf(sysCmd, "echo \"result:-2\" > %s", accFile);
+			system(sysCmd);
+			return 0;
+		}
+		/* Don't check the password if password entry is empty (!) */
+		if (!pw->pw_passwd[0]) {
+			sprintf(sysCmd, "echo \"result:-2\" > %s", accFile);
+			system(sysCmd);
+			return 0;
+		}
+		correct = pw->pw_passwd;
+#if ENABLE_FEATURE_SHADOWPASSWDS
+		if ((correct[0] == 'x' || correct[0] == '*') && !correct[1]) {
+			/* getspnam_r may return 0 yet set result to NULL.
+			 * At least glibc 2.4 does this. Be extra paranoid here. */
+			struct spwd *result = NULL;
+			r = getspnam_r(pw->pw_name, &spw, buffer, sizeof(buffer), &result);
+			correct = (r || !result) ? "aa" : result->sp_pwdp;
+		}
+#endif
+		if (!correct[0]) {/* empty password field? */
+			sprintf(sysCmd, "echo \"result:-2\" > %s", accFile);
+			system(sysCmd);
+			return 0;
+		}
+		encrypted = pw_encrypt(accPwd, correct, 1);
+
+		if (strcmp(encrypted, correct) == 0) {
+			sprintf(sysCmd, "echo \"result:0\" > %s", accFile);
+			system(sysCmd);
+			return 0;
+		}
+		sprintf(sysCmd, "echo \"result:-1\" > %s", accFile);
+		system(sysCmd);
+		return 0;
+	}
+#endif
+//SHA512_PASSWD
 	if (opt & LOGIN_OPT_f) {
 		if (!run_by_root)
 			bb_error_msg_and_die("-f is for root only");
@@ -732,16 +857,24 @@
 
 		if (pw->pw_passwd[0] == '!' || pw->pw_passwd[0] == '*')
 			goto auth_failed;
-
+//SHA512_PASSWD
+#if 0
 		if (opt & LOGIN_OPT_f)
 			break; /* -f USER: success without asking passwd */
-
-		if (pw->pw_uid == 0 && !is_tty_secure(short_tty))
+#endif
+//SHA512_PASSWD
+		if (pw->pw_uid == 0 && !check_securetty(short_tty))
 			goto auth_failed;
 
 		/* Don't check the password if password entry is empty (!) */
 		if (!pw->pw_passwd[0])
+//SHA512_PASSWD
+#if 0
 			break;
+#else
+			goto auth_failed;
+#endif
+//SHA512_PASSWD
  fake_it:
 		/* Password reading and authorization takes place here.
 		 * Note that reads (in no-echo mode) trash tty attributes.
Index: busybox-1.31.1/loginutils/passwd.c
===================================================================
--- busybox-1.31.1.orig/loginutils/passwd.c	2019-06-10 18:50:53.000000000 +0800
+++ busybox-1.31.1/loginutils/passwd.c	2021-03-22 19:46:12.328373928 +0800
@@ -40,7 +40,13 @@
 #include "libbb.h"
 #include <syslog.h>
 
+//SHA512_PASSWD
+#if 0
 static char* new_password(const struct passwd *pw, uid_t myuid, const char *algo)
+#else
+static char* new_password(const struct passwd *pw, uid_t myuid, const char *algo, const char *pwd)
+#endif
+//SHA512_PASSWD
 {
 	char salt[MAX_PW_SALT_LEN];
 	char *orig = NULL;
@@ -64,7 +70,15 @@
 		if (ENABLE_FEATURE_CLEAN_UP)
 			free(encrypted);
 	}
+//SHA512_PASSWD
+#if 0
 	newp = bb_ask_noecho_stdin("New password: "); /* returns malloced str */
+#else
+	if (pwd == NULL)
+		newp = bb_ask_noecho_stdin("New password: "); /* returns malloced str */
+	else newp = (char *)pwd;
+#endif
+//SHA512_PASSWD
 	if (!newp)
 		goto err_ret;
 	if (ENABLE_FEATURE_PASSWD_WEAK_CHECK
@@ -73,7 +87,8 @@
 	) {
 		goto err_ret; /* non-root is not allowed to have weak passwd */
 	}
-
+//SHA512_PASSWD
+#if 0
 	cp = bb_ask_noecho_stdin("Retype password: ");
 	if (!cp)
 		goto err_ret;
@@ -81,7 +96,18 @@
 		puts("Passwords don't match");
 		goto err_ret;
 	}
-
+#else
+	if (pwd == NULL) {
+		cp = bb_ask_noecho_stdin("Retype password: ");
+		if (!cp)
+			goto err_ret;
+		if (strcmp(cp, newp) != 0) {
+			puts("Passwords don't match");
+			goto err_ret;
+		}
+	}
+#endif
+//SHA512_PASSWD
 	crypt_make_pw_salt(salt, algo);
 
 	/* pw_encrypt returns malloced str */
@@ -108,11 +134,17 @@
 		OPT_lock   = (1 << 1), /* -l - lock account */
 		OPT_unlock = (1 << 2), /* -u - unlock account */
 		OPT_delete = (1 << 3), /* -d - delete password */
+//SHA512_PASSWD
+		OPT_pwd = (1 << 4), /* -p - new password */
+//SHA512_PASSWD
 		OPT_lud    = OPT_lock | OPT_unlock | OPT_delete,
 	};
 	unsigned opt;
 	int rc;
 	const char *opt_a = CONFIG_FEATURE_DEFAULT_PASSWD_ALGO;
+//SHA512_PASSWD
+	const char *opt_p = NULL;
+//SHA512_PASSWD
 	const char *filename;
 	char *myname;
 	char *name;
@@ -129,7 +161,13 @@
 
 	logmode = LOGMODE_BOTH;
 	openlog(applet_name, 0, LOG_AUTH);
+//SHA512_PASSWD
+#if 0
 	opt = getopt32(argv, "a:lud", &opt_a);
+#else
+	opt = getopt32(argv, "a:ludp:", &opt_a, &opt_p);
+#endif
+//SHA512_PASSWD
 	//argc -= optind;
 	argv += optind;
 
@@ -183,7 +221,13 @@
 					"locked password for %s", name);
 		}
 		printf("Changing password for %s\n", name);
+//SHA512_PASSWD
+#if 0
 		newp = new_password(pw, myuid, opt_a);
+#else
+		newp = new_password(pw, myuid, opt_a, opt_p);
+#endif
+//SHA512_PASSWD
 		if (!newp) {
 			logmode = LOGMODE_STDIO;
 			bb_error_msg_and_die("password for %s is unchanged", name);
