Index: busybox-1.31.1/networking/traceroute.c
===================================================================
--- busybox-1.31.1.orig/networking/traceroute.c	2021-03-18 10:26:05.354541148 +0800
+++ busybox-1.31.1/networking/traceroute.c	2021-03-18 10:26:05.430541149 +0800
@@ -969,6 +969,7 @@
 	struct sockaddr *to;
 	char *outputrdm_str = NULL;
 #ifdef ZCFG_SUPPORT
+	unsigned int lastRTT = 0;
 	//struct json_object *tracert_msg = NULL, *rt_hops = NULL, *result = NULL;
 	struct json_object *rt_hops = NULL, *result = NULL;
 	const char *payload = NULL;
@@ -1268,6 +1269,7 @@
 		int got_there = 0;
 #ifdef ZCFG_SUPPORT
 		char rtt[272] = {0};
+		char *last_addr, *dst_addr;
 
 		result = json_object_new_object();
 		json_object_array_add(rt_hops, result);
@@ -1320,6 +1322,12 @@
 				}
 
 #ifdef ZCFG_SUPPORT
+				last_addr = xmalloc_sockaddr2dotted_noport(lastaddr);
+				dst_addr = xmalloc_sockaddr2dotted_noport(&dest_lsa->u.sa);
+				if (!strcmp(last_addr, dst_addr))
+					/* last address is the destination address */
+					lastRTT = (t2-t1)/1000;
+
 				print_delta_ms(t1, t2, rtt);
 				if (icmp_code != -1)
 					json_object_object_add(result, "ErrorCode", json_object_new_int(icmp_code));
@@ -1448,6 +1456,7 @@
 	else
 		json_object_object_add(tracert_msg, "DiagnosticsState", json_object_new_string(DIAG_COMPLETE));
 
+	json_object_object_add(tracert_msg, "ResponseTime", json_object_new_int(lastRTT));
 	payload = json_object_to_json_string(tracert_msg);
 
 	payloadLen = strlen(payload) + 1;
