Index: busybox-1.31.1/init/init.c
===================================================================
--- busybox-1.31.1.orig/init/init.c
+++ busybox-1.31.1/init/init.c
@@ -134,6 +134,10 @@
 # include <linux/vt.h>
 # include <sys/sysinfo.h>
 #endif
+#ifdef PLATFORM_ECONET
+#define _GNU_SOURCE
+#include <sched.h>
+#endif
 #include "reboot.h" /* reboot() constants */
 
 #ifdef CONFIG_TT_CUSTOMIZATION_TR069
